R/solve.R

```diff
@@ -0,0 +1,31 @@
+#' @include internal
+NULL
+
+#' Solve a restoration optimization problem
+#'
+#' Solve a restoration optimization problem to generate a solution.
+#'
+# @inheritParams add_max_mesh_objective
+#'
+#' @return A [terra::rast()] object.
+#'
+#' @examples
+#' # TODO
+#'
+#' @export
```

Comment:
Ok so that would be the generic solving function, with the four basic steps: compile, solve using java, prepare the results and return them. It seems perfect, I just wonder why calling it `solve.RestoptProblem`, instead of using something like `restoptr::solve(problem`, it is an R naming convention ?

Reply:
Yeah - that's I was thinking. Awesome! Yeah, method dispatch is a bit more compliated when there are functions with the same name already present in base R packages (or in other commonly used R packages). Since the base R package already contains a `solve()` function, we can't just assign a `solve` function here because it would clash - we need to implement the function in a way that plays nicely with base R. In particular, base R solve function (https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/solve) provides an S3 generic. This means that if we create a `solve.RestoptProblem` function then R will automatically use this `solve.RestoptProblem` function if the user runs `solve(x)` where `x` is a `RestoptProblem` class object. In other words, S3 methods are a way to dispatch functions/methods using a `${method_name}.${object_class}` style suffix system based on the first argument/parameter (S4 methods are handy because they let you dispatch on multiple parameters/arguments). Does that make sense? Sorry, I'm probably getting some of this terminology wrong. If it's helpful, here's some info on S3 objects and methods: http://adv-r.had.co.nz/S3.html

Reply:
It is perfectly clear :-)

Reply:
Awesome!